### PR TITLE
[Model Runner V2] Support thinking_token_budget for mrv2

### PR DESCRIPTION
**Pull request by xhdidi**
## Purpose
Supported limiting the maximum length of "reasoning" via "thinking_token_budget" in mrv2.

## Test Plan
`curl http://localhost:8000/v1/chat/completions     -H "Content-Type: application/json"     -d '{
        "model": "Qwen3.5-35B-A3B",
        "messages": [
            {"role": "system", "content": "You are a helpful assistant."},
            {"role": "user", "content": "How's the weather today?"}
        ],
         "ignore_eos": true,
         "max_tokens": 100,
         "thinking_token_budget": 50
    }'
`

### CODE REVIEW

**Review by chaunceyjiang — COMMENTED** [review]
I’m more inclined toward https://github.com/vllm-project/vllm/pull/46727, which may be more efficient.

**Comment by ma-leezc** [comment]
This PR looks stalled in favour of #46727, but #46727 landed **without** the
speculative-decoding half, so the gap this PR closes is still open on main. The
symptom is #39573.

**What is missing on main**

`ThinkingBudgetState.apply()` is only reached from `Sampler.__call__`. The model
runner routes past it whenever there are draft tokens:

```python
elif input_batch.num_draft_tokens == 0 or self.rejection_sampler is None:
    sampler_output = self.sampler(logits, input_batch)     # budget applied
else:
    # Rejection sampling for spec decoding.
    sampler_output = self.rejection_sampler(
        logits, input_batch, self.speculator.draft_logits,  # budget never applied
    )
```

So with any speculative decoding enabled under Model Runner V2,
`thinking_token_budget` has no effect at all. `vllm/v1/worker/gpu/spec_decode/
rejection_sampler.py` has no reference to the budget state, and this PR's change
there — threading `input_batch` into `apply_sampling_params` — is what makes it
reachable.

V1 does not have this gap. `vllm/v1/sample/rejection_sampler.py` applies the
budget explicitly during verification, and passes the draft tokens in:

```python
holder = sampling_metadata.thinking_budget_state_holder
if holder is not None and holder.has_tracked_requests():
    logits = holder.apply_to_logits(
        logits,
        predict_bonus_token=False,
        spec_token_ids=sampling_metadata.spec_token_ids,
    )
```

**Measured**

Same build, same model, same request; MTP left enabled throughout
(`{"method": "mtp", "num_speculative_tokens": 4}`), `max_tokens: 2000`. The only
variable is `VLLM_USE_V2_MODEL_RUNNER`:

| runner | `thinking_token_budget` | `reasoning_tokens` | `finish_reason` | content |
|---|---|---|---|---|
| V2 | unset | 2000 | length | empty |
| V2 | 64 | 2000 | length | empty |
| V2 | 200 | 2000 | length | empty |
| V1 | unset | 2000 | length | empty |
| V1 | **64** | **63** | stop | 1007 chars |
| V1 | **200** | **199** | stop | 1707 chars |

The `unset` rows are the control and match across runners, so this isolates to
the runner selection.

Worth noting the failure mode is worse than an ignored parameter: reasoning
consumes the whole `max_tokens` allowance and the response carries **no answer
at all**. There is no server-side setting that bounds it — `enable_thinking` is
all-or-nothing — so on V2 the only workaround is disabling speculative decoding
or setting `VLLM_USE_V2_MODEL_RUNNER=0`. Since v0.29.0 makes V2 the default for
all models, that workaround is getting narrower.

I also tried #52677 (applies cleanly to a 44fe2a392-based build) and a
multi-token `reasoning_end_str` so that `track_forced_end` becomes true; neither
changed the V2 behaviour, which is consistent with the routing above — the
forced-end machinery is never consulted on the rejection path.

Happy to rebase-test this PR and report back if that is useful.

**Comment by ma-leezc** [comment]
Correction to my comment above — I got the mechanism wrong, and I want to fix
that before anyone acts on it.

I claimed the budget is never applied on the rejection path. That is not true.
`apply_sampling_params` is shared by both paths and already contains the budget
apply, and the rejection sampler does call it:

```
RejectionSampler.__call__ -> _verify_in_chunks -> _verify
  -> self.sampler.apply_sampling_params(...)     # spec_decode/rejection_sampler.py
     -> self.thinking_budget_state.apply(...)    # sample/sampler.py
```

`needs_logits_processing` is also set for budget requests, so the early return
at the top of `apply_sampling_params` does not skip them. I have separately
observed `_thinking_budget_kernel` being JIT-compiled during inference on a V2
run with a budget set, which confirms the kernel really does execute there.

So the forcing runs on the verification path and the output is still unaffected.

What does differ between the two runners is speculative-decoding *awareness*:

- V1: `holder.apply_to_logits(logits, predict_bonus_token=False, spec_token_ids=...)`
- V2: `ThinkingBudgetState.apply(logits, expanded_idx_mapping, idx_mapping,
  idx_mapping_np, input_ids, expanded_local_pos)`

V1's entry point is given the draft tokens and a bonus-token flag. V2's takes no
speculative parameters at all, so it is applied while verifying drafts without
knowing that is what it is doing.

I do not have a verified explanation for why that produces *zero* observable
effect rather than a partial one, so I will not guess again here. The
measurements in my previous comment stand — they were direct observations, and
only the mechanism I inferred from them was wrong:

- V2 + MTP: budgets 64 and 200 both yield 2000 reasoning tokens and empty content
- V1 + MTP: 63 and 199 reasoning tokens with real answers
- controls (no budget) identical on both

The related note in #52677 — that a forced end token can be rejected under
speculative decoding, and that the budget machinery's rejected-end recovery then
flips the request back — looks like the more promising direction, but I have not
verified it.

Apologies for the noise.

**Comment by njhill** [comment]
Closing as dup of https://github.com/vllm-project/vllm/pull/46727.
